File: windows/src/Core/DiagnosticLog.cs
#nullable enable
using System;
using System.Diagnostics;
using System.IO;
using System.Reflection;
using System.Runtime.InteropServices;
using System.Text;
using System.Text.RegularExpressions;

namespace YTray.Core
{
    /// <summary>
    /// Small, dependency-free application logger. It deliberately owns no long-lived file handle,
    /// so support staff can read, copy, archive, or delete logs while YTray is running.
    /// </summary>
    internal static class DiagnosticLog
    {
        private const long MainLogLimit = 5L * 1024 * 1024;
        private const long ErrorLogLimit = 2L * 1024 * 1024;
        private const int BackupCount = 3;
        private const int MaximumEntryLength = 64 * 1024;
        private static readonly object SyncRoot = new object();
        private static readonly UTF8Encoding Utf8WithoutBom = new UTF8Encoding(false);
        private static readonly Regex UrlCredentials = new Regex(
            @"(?<scheme>[a-z][a-z0-9+.-]*://)[^\s/@:]+:[^\s/@]+@",
            RegexOptions.IgnoreCase | RegexOptions.Compiled);
        private static readonly Regex AuthorizationValues = new Regex(
            @"(?<key>proxy-authorization|authorization)(?<separator>\s*[:=]\s*)(?<value>[^\r\n]+)",
            RegexOptions.IgnoreCase | RegexOptions.Compiled);
        private static readonly Regex SecretValues = new Regex(
            @"(?<key>password|passwd|pwd|token|secret)(?<keyquote>[\""']?)(?<separator>\s*[:=]\s*)(?<valuequote>[\""']?)(?<value>[^\s,;\]\}\""']+)(?<endquote>[\""']?)",
            RegexOptions.IgnoreCase | RegexOptions.Compiled);

        private static string? _logDirectory;
        private static string? _userProfile;
        private static bool _initialized;

        internal static string LogDirectory => _logDirectory
            ?? Path.Combine(StatePersistence.DefaultApplicationDirectory, "Logs");
        internal static string MainLogPath => Path.Combine(LogDirectory, "ytray.log");
        internal static string ErrorLogPath => Path.Combine(LogDirectory, "ytray-errors.log");

        internal static void Initialize(string applicationDirectory)
        {
            if (string.IsNullOrWhiteSpace(applicationDirectory))
                throw new ArgumentException("Application directory is required.", nameof(applicationDirectory));

            lock (SyncRoot)
            {
                var directory = Path.Combine(applicationDirectory, "Logs");
                if (_initialized && string.Equals(_logDirectory, directory, StringComparison.OrdinalIgnoreCase)) return;

                _logDirectory = directory;
                _userProfile = Environment.GetFolderPath(Environment.SpecialFolder.UserProfile)
                    .TrimEnd(Path.DirectorySeparatorChar, Path.AltDirectorySeparatorChar);
                try
                {
                    Directory.CreateDirectory(directory);
                    RotateIfNeeded(MainLogPath, MainLogLimit, BackupCount);
                    RotateIfNeeded(ErrorLogPath, ErrorLogLimit, BackupCount);
                    Touch(MainLogPath);
                    Touch(ErrorLogPath);
                    StatePersistence.RestrictToCurrentUser(MainLogPath);
                    StatePersistence.RestrictToCurrentUser(ErrorLogPath);
                    _initialized = true;
                    WriteCore(MainLogPath, "INFO", "app.start",
                        $"session started; version={ApplicationVersion()}; os={Environment.OSVersion.VersionString}; " +
                        $"processArchitecture={ProcessArchitecture()}; is64BitOS={Environment.Is64BitOperatingSystem}");
                }
                catch
                {
                    // Diagnostics must never make application startup fail.
                    _initialized = false;
                }
            }
        }

        internal static void Info(string category, string message) =>
            Write("INFO", category, message, false);

        internal static void Warning(string category, string message) =>
            Write("WARN", category, message, false);

        internal static void Error(string category, Exception exception, string? message = null)
        {
            if (exception == null) return;
            var detail = string.IsNullOrWhiteSpace(message)
                ? exception.ToString()
                : message + Environment.NewLine + exception;
            Write("ERROR", category, detail, true);
        }

        internal static bool TryOpen(out string openedPath, out string? error)
        {
            openedPath = MainLogPath;
            error = null;
            try
            {
                if (!_initialized) Initialize(StatePersistence.DefaultApplicationDirectory);
                Touch(MainLogPath);
                var process = Process.Start(new ProcessStartInfo
                {
                    FileName = "notepad.exe",
                    Arguments = "\"" + MainLogPath.Replace("\"", "") + "\"",
                    UseShellExecute = true,
                });
                if (process == null) throw new InvalidOperationException("Notepad did not start.");
                Info("diagnostics.open", "opened main diagnostic log");
                return true;
            }
            catch (Exception primary)
            {
                try
                {
                    Directory.CreateDirectory(LogDirectory);
                    Process.Start(new ProcessStartInfo
                    {
                        FileName = LogDirectory,
                        UseShellExecute = true,
                    });
                    openedPath = LogDirectory;
                    Warning("diagnostics.open", "opening the log file failed; opened the log directory instead");
                    return true;
                }
                catch (Exception fallback)
                {
                    error = "无法打开诊断日志：" + fallback.Message;
                    Error("diagnostics.open", fallback, primary.Message);
                    return false;
                }
            }
        }

        private static void Write(string level, string category, string message, bool alsoErrorLog)
        {
            if (!_initialized) return;
            lock (SyncRoot)
            {
                try
                {
                    RotateIfNeeded(MainLogPath, MainLogLimit, BackupCount);
                    WriteCore(MainLogPath, level, category, message);
                    if (alsoErrorLog)
                    {
                        RotateIfNeeded(ErrorLogPath, ErrorLogLimit, BackupCount);
                        WriteCore(ErrorLogPath, level, category, message);
                    }
                }
                catch
                {
                    // Logging failures are intentionally swallowed to avoid recursive crashes.
                }
            }
        }

        private static void WriteCore(string path, string level, string category, string message)
        {
            var safeCategory = Sanitize(category).Replace("\r", " ").Replace("\n", " ");
            var safeMessage = Sanitize(message);
            if (safeMessage.Length > MaximumEntryLength)
                safeMessage = safeMessage.Substring(0, MaximumEntryLength) + Environment.NewLine + "[entry truncated]";
            var entry = $"[{DateTimeOffset.Now:O}] [{level}] [{safeCategory}] [pid:{Process.GetCurrentProcess().Id}] {safeMessage}{Environment.NewLine}";
            using (var stream = new FileStream(path, FileMode.Append, FileAccess.Write,
                       FileShare.ReadWrite | FileShare.Delete, 4096, FileOptions.SequentialScan))
            using (var writer = new StreamWriter(stream, Utf8WithoutBom))
                writer.Write(entry);
        }

        internal static string Sanitize(string? value)
        {
            var result = value ?? "";
            result = UrlCredentials.Replace(result, "${scheme}***:***@");
            result = AuthorizationValues.Replace(result, "${key}${separator}***");
            result = SecretValues.Replace(result,
                "${key}${keyquote}${separator}${valuequote}***${endquote}");
            var userProfile = _userProfile;
            if (!string.IsNullOrWhiteSpace(userProfile))
                result = ReplaceOrdinalIgnoreCase(result, userProfile!, "%USERPROFILE%");
            return result;
        }

        private static string ReplaceOrdinalIgnoreCase(string source, string oldValue, string newValue)
        {
            var offset = 0;
            var index = source.IndexOf(oldValue, offset, StringComparison.OrdinalIgnoreCase);
            if (index < 0) return source;
            var builder = new StringBuilder(source.Length);
            while (index >= 0)
            {
                builder.Append(source, offset, index - offset).Append(newValue);
                offset = index + oldValue.Length;
                index = source.IndexOf(oldValue, offset, StringComparison.OrdinalIgnoreCase);
            }
            return builder.Append(source, offset, source.Length - offset).ToString();
        }

        private static void Touch(string path)
        {
            Directory.CreateDirectory(Path.GetDirectoryName(path)!);
            using (new FileStream(path, FileMode.Append, FileAccess.Write,
                       FileShare.ReadWrite | FileShare.Delete)) { }
        }

        internal static void RotateIfNeeded(string path, long maximumBytes, int backupCount)
        {
            if (maximumBytes <= 0 || backupCount <= 0 || !File.Exists(path)) return;
            var file = new FileInfo(path);
            if (file.Length < maximumBytes) return;
            for (var index = backupCount; index >= 1; index--)
            {
                var destination = BackupPath(path, index);
                if (index == backupCount && File.Exists(destination)) File.Delete(destination);
                var source = index == 1 ? path : BackupPath(path, index - 1);
                if (File.Exists(source)) File.Move(source, destination);
            }
        }

        private static string BackupPath(string path, int index) =>
            Path.Combine(Path.GetDirectoryName(path)!,
                Path.GetFileNameWithoutExtension(path) + "." + index + Path.GetExtension(path));

        private static string ApplicationVersion()
        {
            var informational = typeof(DiagnosticLog).Assembly
                .GetCustomAttribute<AssemblyInformationalVersionAttribute>()?.InformationalVersion;
            return string.IsNullOrWhiteSpace(informational)
                ? typeof(DiagnosticLog).Assembly.GetName().Version?.ToString() ?? "unknown"
                : informational!;
        }

        private static string ProcessArchitecture()
        {
            try { return RuntimeInformation.ProcessArchitecture.ToString(); }
            catch { return Environment.Is64BitProcess ? "x64" : "x86"; }
        }

        internal static void ResetForTests()
        {
            lock (SyncRoot)
            {
                _initialized = false;
                _logDirectory = null;
                _userProfile = null;
            }
        }
    }
}

File: darwin/Sources/YTray/DiagnosticLog.swift
import AppKit
import Foundation

/// Persistent support diagnostics shared by the macOS application workflows. The logger keeps no
/// long-lived file handle, so logs remain readable and copyable while YTray is running.
enum DiagnosticLog {
    private static let lock = NSLock()
    private static let mainLogLimit: UInt64 = 5 * 1024 * 1024
    private static let errorLogLimit: UInt64 = 2 * 1024 * 1024
    private static let backupCount = 3
    private static let maximumEntryCharacters = 64 * 1024
    private static var applicationDirectory: URL?
    private static var initialized = false

    static var logDirectory: URL {
        let base = applicationDirectory ?? AppEnvironment.defaultApplicationDirectory
        return base.appendingPathComponent("Logs", isDirectory: true)
    }

    static var mainLogURL: URL { logDirectory.appendingPathComponent("ytray.log") }
    static var errorLogURL: URL { logDirectory.appendingPathComponent("ytray-errors.log") }

    static func initialize(applicationDirectory: URL) {
        lock.lock()
        defer { lock.unlock() }
        let normalized = applicationDirectory.standardizedFileURL
        if initialized, self.applicationDirectory?.standardizedFileURL == normalized { return }
        self.applicationDirectory = normalized
        do {
            try FileManager.default.createDirectory(
                at: logDirectory,
                withIntermediateDirectories: true
            )
            try rotateIfNeeded(mainLogURL, maximumBytes: mainLogLimit, backupCount: backupCount)
            try rotateIfNeeded(errorLogURL, maximumBytes: errorLogLimit, backupCount: backupCount)
            try touch(mainLogURL)
            try touch(errorLogURL)
            initialized = true
            try writeCore(
                to: mainLogURL,
                level: "INFO",
                category: "app.start",
                message: "session started; version=\(applicationVersion); os=\(ProcessInfo.processInfo.operatingSystemVersionString); architecture=\(architecture)"
            )
        } catch {
            // Diagnostics must never prevent the application from starting.
            initialized = false
        }
    }

    static func info(_ category: String, _ message: String) {
        write(level: "INFO", category: category, message: message, alsoToErrorLog: false)
    }

    static func warning(_ category: String, _ message: String) {
        write(level: "WARN", category: category, message: message, alsoToErrorLog: false)
    }

    static func error(_ category: String, _ error: Error, message: String? = nil) {
        let heading = message.map { $0 + "\n" } ?? ""
        let detail = "\(heading)\(String(reflecting: type(of: error))): \(error.localizedDescription)\n"
            + Thread.callStackSymbols.joined(separator: "\n")
        write(level: "ERROR", category: category, message: detail, alsoToErrorLog: true)
    }

    @MainActor
    static func openMainLog(applicationDirectory: URL) throws {
        initialize(applicationDirectory: applicationDirectory)
        try touch(mainLogURL)
        if NSWorkspace.shared.open(mainLogURL) {
            info("diagnostics.open", "opened main diagnostic log")
            return
        }
        if NSWorkspace.shared.open(logDirectory) {
            warning("diagnostics.open", "opening the log file failed; opened the log directory instead")
            return
        }
        throw DiagnosticLogError.openFailed
    }

    private static func write(
        level: String,
        category: String,
        message: String,
        alsoToErrorLog: Bool
    ) {
        lock.lock()
        defer { lock.unlock() }
        guard initialized else { return }
        do {
            try rotateIfNeeded(mainLogURL, maximumBytes: mainLogLimit, backupCount: backupCount)
            try writeCore(to: mainLogURL, level: level, category: category, message: message)
            if alsoToErrorLog {
                try rotateIfNeeded(errorLogURL, maximumBytes: errorLogLimit, backupCount: backupCount)
                try writeCore(to: errorLogURL, level: level, category: category, message: message)
            }
        } catch {
            // Logging failures are intentionally swallowed to avoid recursive failures.
        }
    }

    private static func writeCore(to url: URL, level: String, category: String, message: String) throws {
        let safeCategory = sanitize(category)
            .replacingOccurrences(of: "\r", with: " ")
            .replacingOccurrences(of: "\n", with: " ")
        var safeMessage = sanitize(message)
        if safeMessage.count > maximumEntryCharacters {
            safeMessage = String(safeMessage.prefix(maximumEntryCharacters)) + "\n[entry truncated]"
        }
        let formatter = ISO8601DateFormatter()
        formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        let entry = "[\(formatter.string(from: Date()))] [\(level)] [\(safeCategory)] [pid:\(ProcessInfo.processInfo.processIdentifier)] \(safeMessage)\n"
        let handle = try FileHandle(forWritingTo: url)
        defer { try? handle.close() }
        try handle.seekToEnd()
        try handle.write(contentsOf: Data(entry.utf8))
    }

    static func sanitize(_ value: String) -> String {
        var result = value
        result = replacing(
            in: result,
            pattern: #"(?i)([a-z][a-z0-9+.-]*://)[^\s/@:]+:[^\s/@]+@"#,
            with: "$1***:***@"
        )
        result = replacing(
            in: result,
            pattern: #"(?im)(proxy-authorization|authorization)(\s*[:=]\s*)[^\r\n]+"#,
            with: "$1$2***"
        )
        result = replacing(
            in: result,
            pattern: #"(?i)(password|passwd|pwd|token|secret)(["']?)(\s*[:=]\s*)(["']?)[^\s,;\]\}"']+(["']?)"#,
            with: "$1$2$3$4***$5"
        )
        let home = FileManager.default.homeDirectoryForCurrentUser.path
        if !home.isEmpty {
            result = replacing(
                in: result,
                pattern: NSRegularExpression.escapedPattern(for: home),
                with: "%USERPROFILE%",
                options: [.caseInsensitive]
            )
        }
        return result
    }

    private static func replacing(
        in value: String,
        pattern: String,
        with replacement: String,
        options: NSRegularExpression.Options = []
    ) -> String {
        guard let expression = try? NSRegularExpression(pattern: pattern, options: options) else { return value }
        return expression.stringByReplacingMatches(
            in: value,
            range: NSRange(value.startIndex..., in: value),
            withTemplate: replacement
        )
    }

    private static func touch(_ url: URL) throws {
        try FileManager.default.createDirectory(
            at: url.deletingLastPathComponent(),
            withIntermediateDirectories: true
        )
        if !FileManager.default.fileExists(atPath: url.path) {
            guard FileManager.default.createFile(atPath: url.path, contents: nil) else {
                throw CocoaError(.fileWriteUnknown)
            }
        }
        try? FileManager.default.setAttributes(
            [.posixPermissions: 0o600],
            ofItemAtPath: url.path
        )
    }

    static func rotateIfNeeded(_ url: URL, maximumBytes: UInt64, backupCount: Int) throws {
        guard maximumBytes > 0, backupCount > 0,
              let size = try? url.resourceValues(forKeys: [.fileSizeKey]).fileSize,
              UInt64(size) >= maximumBytes else { return }
        let fileManager = FileManager.default
        for index in stride(from: backupCount, through: 1, by: -1) {
            let destination = backupURL(for: url, index: index)
            if index == backupCount { try? fileManager.removeItem(at: destination) }
            let source = index == 1 ? url : backupURL(for: url, index: index - 1)
            if fileManager.fileExists(atPath: source.path) {
                try fileManager.moveItem(at: source, to: destination)
            }
        }
    }

    private static func backupURL(for url: URL, index: Int) -> URL {
        let stem = url.deletingPathExtension().lastPathComponent
        let ext = url.pathExtension
        return url.deletingLastPathComponent()
            .appendingPathComponent("\(stem).\(index).\(ext)")
    }

    private static var applicationVersion: String {
        Bundle.main.object(forInfoDictionaryKey: "CFBundleShortVersionString") as? String ?? "development"
    }

    private static var architecture: String {
        #if arch(arm64)
        return "arm64"
        #elseif arch(x86_64)
        return "x86_64"
        #else
        return "unknown"
        #endif
    }

    static func resetForTesting() {
        lock.lock()
        defer { lock.unlock() }
        initialized = false
        applicationDirectory = nil
    }
}

private enum DiagnosticLogError: LocalizedError {
    case openFailed

    var errorDescription: String? {
        "无法打开诊断日志"
    }
}
